File: test/reference/parser.ts
// The TypeScript tape parser as it stood before the Rust core replaced it, kept
// as the reference implementation the core is tested against (test/core.test.ts),
// the way the audio tests keep the original sample sink. It is not part of the
// app bundle and must not gain features: it is the frozen second opinion.
//
// `scripts/dump-blocks.mjs` dumps tapes through this parser to generate the
// fixtures of the Rust differential test, so the two implementations stay
// independent of each other.
import { Reader } from '../../src/tzx/bytes';
import { Block, newUid, ParsedTape, SymDef, PilotRun } from '../../src/tzx/types';

const TZX_SIGNATURE = 'ZXTape!\x1a';

export function isTzx(buf: Uint8Array): boolean {
  if (buf.length < 10) return false;
  for (let i = 0; i < 8; i++) if (buf[i] !== TZX_SIGNATURE.charCodeAt(i)) return false;
  return true;
}

export function parseTzx(buf: Uint8Array): ParsedTape {
  if (!isTzx(buf)) throw new Error('Not a TZX file (missing ZXTape! signature)');
  const r = new Reader(buf);
  r.pos = 8;
  const major = r.u8();
  const minor = r.u8();
  const blocks: Block[] = [];
  const warnings: string[] = [];
  while (!r.eof()) {
    const start = r.pos;
    const id = r.u8();
    try {
      blocks.push(parseBlock(id, r));
    } catch (e) {
      warnings.push(`Block ${blocks.length + 1} (ID ${id.toString(16).padStart(2, '0')}) at offset ${start}: ${(e as Error).message}`);
      // Keep whatever is left as an unknown block so nothing is silently lost.
      r.pos = start + 1;
      blocks.push({ uid: newUid(), id, unknown: true, raw: r.bytes(r.remaining) });
      break;
    }
  }
  return { blocks, major, minor, warnings };
}

function readSymDefs(r: Reader, count: number, maxPulses: number): SymDef[] {
  const out: SymDef[] = [];
  for (let i = 0; i < count; i++) {
    const flags = r.u8();
    const pulses: number[] = [];
    for (let p = 0; p < maxPulses; p++) pulses.push(r.u16());
    // trim trailing zero pulses
    while (pulses.length > 0 && pulses[pulses.length - 1] === 0) pulses.pop();
    out.push({ flags, pulses });
  }
  return out;
}

export function bitsPerSymbol(alphabetSize: number): number {
  if (alphabetSize <= 1) return 1;
  return Math.ceil(Math.log2(alphabetSize));
}

export function parseBlock(id: number, r: Reader): Block {
  const uid = newUid();
  switch (id) {
    case 0x10: {
      const pause = r.u16();
      const len = r.u16();
      return { uid, id, pause, data: r.bytes(len) };
    }
    case 0x11: {
      const pilot = r.u16();
      const sync1 = r.u16();
      const sync2 = r.u16();
      const zero = r.u16();
      const one = r.u16();
      const pilotLen = r.u16();
      const usedBits = r.u8();
      const pause = r.u16();
      const len = r.u24();
      return { uid, id, pilot, sync1, sync2, zero, one, pilotLen, usedBits, pause, data: r.bytes(len) };
    }
    case 0x12: {
      const pulseLen = r.u16();
      const count = r.u16();
      return { uid, id, pulseLen, count };
    }
    case 0x13: {
      const n = r.u8();
      const pulses: number[] = [];
      for (let i = 0; i < n; i++) pulses.push(r.u16());
      return { uid, id, pulses };
    }
    case 0x14: {
      const zero = r.u16();
      const one = r.u16();
      const usedBits = r.u8();
      const pause = r.u16();
      const len = r.u24();
      return { uid, id, zero, one, usedBits, pause, data: r.bytes(len) };
    }
    case 0x15: {
      const tstates = r.u16();
      const pause = r.u16();
      const usedBits = r.u8();
      const len = r.u24();
      return { uid, id, tstates, pause, usedBits, data: r.bytes(len) };
    }
    case 0x18: {
      const len = r.u32();
      // A length under the 10-byte header is corrupt; reading len - 10 bytes would
      // move the read position backwards. Report it and keep the rest verbatim.
      if (len < 10) throw new Error(`CSW block length ${len} is shorter than its 10-byte header`);
      const pause = r.u16();
      const sampleRate = r.u24();
      const compression = r.u8();
      const pulseCount = r.u32();
      return { uid, id, pause, sampleRate, compression, pulseCount, data: r.bytes(len - 10) };
    }
    case 0x19: {
      const len = r.u32();
      const end = r.pos + len;
      const pause = r.u16();
      const totp = r.u32();
      const npp = r.u8();
      let asp = r.u8();
      const totd = r.u32();
      const npd = r.u8();
      let asd = r.u8();
      if (asp === 0) asp = 256;
      if (asd === 0) asd = 256;
      let pilotSymbols: SymDef[] = [];
      let pilotStream: PilotRun[] = [];
      if (totp > 0) {
        pilotSymbols = readSymDefs(r, asp, npp);
        for (let i = 0; i < totp; i++) {
          const symbol = r.u8();
          const reps = r.u16();
          pilotStream.push({ symbol, reps });
        }
      }
      let dataSymbols: SymDef[] = [];
      let data = new Uint8Array(0);
      if (totd > 0) {
        dataSymbols = readSymDefs(r, asd, npd);
        const ds = Math.ceil((bitsPerSymbol(asd) * totd) / 8);
        data = r.bytes(ds);
      }
      if (r.pos !== end) {
        // Some files are sloppy; trust the declared block length.
        r.pos = end;
      }
      return { uid, id, pause, totp, npp, pilotSymbols, pilotStream, totd, npd, dataSymbols, data };
    }
    case 0x20:
      return { uid, id, pause: r.u16() };
    case 0x21: {
      const n = r.u8();
      return { uid, id, name: r.str(n) };
    }
    case 0x22:
      return { uid, id };
    case 0x23:
      return { uid, id, offset: r.i16() };
    case 0x24:
      return { uid, id, count: r.u16() };
    case 0x25:
      return { uid, id };
    case 0x26: {
      const n = r.u16();
      const offsets: number[] = [];
      for (let i = 0; i < n; i++) offsets.push(r.i16());
      return { uid, id, offsets };
    }
    case 0x27:
      return { uid, id };
    case 0x28: {
      const len = r.u16();
      const end = r.pos + len;
      const n = r.u8();
      const entries = [];
      for (let i = 0; i < n; i++) {
        const offset = r.i16();
        const l = r.u8();
        entries.push({ offset, text: r.str(l) });
      }
      r.pos = end;
      return { uid, id, entries };
    }
    case 0x2a: {
      r.u32(); // length, always 0
      return { uid, id };
    }
    case 0x2b: {
      r.u32(); // length, always 1
      return { uid, id, level: r.u8() };
    }
    case 0x30: {
      const n = r.u8();
      return { uid, id, text: r.str(n) };
    }
    case 0x31: {
      const time = r.u8();
      const n = r.u8();
      return { uid, id, time, text: r.str(n) };
    }
    case 0x32: {
      const len = r.u16();
      const end = r.pos + len;
      const n = r.u8();
      const entries = [];
      for (let i = 0; i < n; i++) {
        const type = r.u8();
        const l = r.u8();
        entries.push({ type, text: r.str(l) });
      }
      r.pos = end;
      return { uid, id, entries };
    }
    case 0x33: {
      const n = r.u8();
      const entries = [];
      for (let i = 0; i < n; i++) entries.push({ type: r.u8(), id: r.u8(), info: r.u8() });
      return { uid, id, entries };
    }
    case 0x35: {
      const ident = r.str(16);
      const len = r.u32();
      return { uid, id, ident, data: r.bytes(len) };
    }
    case 0x5a:
      return { uid, id, raw: r.bytes(9) };
    // Deprecated blocks: keep raw so they round-trip.
    case 0x16:
    case 0x17: {
      const start = r.pos;
      const len = r.u32();
      r.pos = start;
      return { uid, id, unknown: true, raw: r.bytes(4 + len) };
    }
    case 0x34:
      return { uid, id, unknown: true, raw: r.bytes(8) };
    case 0x40: {
      const start = r.pos;
      r.u8();
      const len = r.u24();
      r.pos = start;
      return { uid, id, unknown: true, raw: r.bytes(4 + len) };
    }
    default: {
      // General extension rule: any unknown block has a DWORD length.
      const start = r.pos;
      const len = r.u32();
      r.pos = start;
      return { uid, id, unknown: true, raw: r.bytes(4 + len) };
    }
  }
}

/** Parse a TAP file into standard speed blocks. */
export function parseTap(buf: Uint8Array): ParsedTape {
  const r = new Reader(buf);
  const blocks: Block[] = [];
  const warnings: string[] = [];
  while (!r.eof()) {
    if (r.remaining < 2) {
      warnings.push('Trailing byte ignored at end of TAP file');
      break;
    }
    const len = r.u16();
    if (len > r.remaining) {
      warnings.push(`Truncated TAP block ${blocks.length + 1}: declared ${len} bytes, ${r.remaining} available`);
      blocks.push({ uid: newUid(), id: 0x10, pause: 1000, data: r.bytes(r.remaining) });
      break;
    }
    blocks.push({ uid: newUid(), id: 0x10, pause: 1000, data: r.bytes(len) });
  }
  return { blocks, major: 1, minor: 20, warnings };
}

/** Auto-detect TZX vs TAP by signature. */
export function parseTape(buf: Uint8Array): ParsedTape {
  return isTzx(buf) ? parseTzx(buf) : parseTap(buf);
}

File: core/src/parser.rs
//! TZX and TAP parsing, the port of `src/tzx/parser.ts`.

use crate::bytes::{ReadError, ReadResult, Reader};
use crate::types::{ArchiveEntry, Block, Body, HardwareEntry, PilotRun, SelectEntry, SymDef};

#[derive(Clone, Debug, Default)]
pub struct ParsedTape {
    pub blocks: Vec<Block>,
    pub major: u8,
    pub minor: u8,
    pub warnings: Vec<String>,
}

const TZX_SIGNATURE: &[u8; 8] = b"ZXTape!\x1a";

pub fn is_tzx(buf: &[u8]) -> bool {
    buf.len() >= 10 && &buf[..8] == TZX_SIGNATURE
}

pub fn parse_tzx(buf: &[u8]) -> Result<ParsedTape, String> {
    if !is_tzx(buf) {
        return Err("Not a TZX file (missing ZXTape! signature)".to_string());
    }
    let mut r = Reader::new(buf);
    r.pos = 8;
    // The signature check guarantees these two.
    let major = r.u8().unwrap();
    let minor = r.u8().unwrap();
    let mut blocks: Vec<Block> = Vec::new();
    let mut warnings: Vec<String> = Vec::new();
    while !r.eof() {
        let start = r.pos;
        let id = r.u8().unwrap();
        match parse_block(id, &mut r) {
            Ok(body) => blocks.push(Block::new(body)),
            Err(e) => {
                warnings.push(format!(
                    "Block {} (ID {:02x}) at offset {}: {}",
                    blocks.len() + 1,
                    id,
                    start,
                    e
                ));
                // Keep whatever is left as an unknown block so nothing is silently lost.
                r.pos = start + 1;
                let raw = r.bytes(r.remaining()).unwrap();
                blocks.push(Block::new(Body::Unknown { id, raw }));
                break;
            }
        }
    }
    Ok(ParsedTape { blocks, major, minor, warnings })
}

fn read_sym_defs(r: &mut Reader, count: usize, max_pulses: u8) -> ReadResult<Vec<SymDef>> {
    let mut out = Vec::with_capacity(count);
    for _ in 0..count {
        let flags = r.u8()?;
        let mut pulses = Vec::with_capacity(max_pulses as usize);
        for _ in 0..max_pulses {
            pulses.push(r.u16()?);
        }
        // trim trailing zero pulses
        while pulses.last() == Some(&0) {
            pulses.pop();
        }
        out.push(SymDef { flags, pulses });
    }
    Ok(out)
}

pub fn bits_per_symbol(alphabet_size: u32) -> u32 {
    if alphabet_size <= 1 {
        return 1;
    }
    u32::BITS - (alphabet_size - 1).leading_zeros()
}

pub fn parse_block(id: u8, r: &mut Reader) -> ReadResult<Body> {
    Ok(match id {
        0x10 => {
            let pause = r.u16()?;
            let len = r.u16()? as usize;
            Body::Standard { pause, data: r.bytes(len)? }
        }
        0x11 => {
            let pilot = r.u16()?;
            let sync1 = r.u16()?;
            let sync2 = r.u16()?;
            let zero = r.u16()?;
            let one = r.u16()?;
            let pilot_len = r.u16()?;
            let used_bits = r.u8()?;
            let pause = r.u16()?;
            let len = r.u24()? as usize;
            Body::Turbo { pilot, sync1, sync2, zero, one, pilot_len, used_bits, pause, data: r.bytes(len)? }
        }
        0x12 => Body::PureTone { pulse_len: r.u16()?, count: r.u16()? },
        0x13 => {
            let n = r.u8()?;
            let mut pulses = Vec::with_capacity(n as usize);
            for _ in 0..n {
                pulses.push(r.u16()?);
            }
            Body::PulseSeq { pulses }
        }
        0x14 => {
            let zero = r.u16()?;
            let one = r.u16()?;
            let used_bits = r.u8()?;
            let pause = r.u16()?;
            let len = r.u24()? as usize;
            Body::PureData { zero, one, used_bits, pause, data: r.bytes(len)? }
        }
        0x15 => {
            let tstates = r.u16()?;
            let pause = r.u16()?;
            let used_bits = r.u8()?;
            let len = r.u24()? as usize;
            Body::Direct { tstates, pause, used_bits, data: r.bytes(len)? }
        }
        0x18 => {
            let len = r.u32()? as usize;
            // A length under the 10-byte header is corrupt; reading len - 10 bytes
            // would move the read position backwards.
            if len < 10 {
                return Err(ReadError(format!(
                    "CSW block length {} is shorter than its 10-byte header",
                    len
                )));
            }
            let pause = r.u16()?;
            let sample_rate = r.u24()?;
            let compression = r.u8()?;
            let pulse_count = r.u32()?;
            Body::Csw { pause, sample_rate, compression, pulse_count, data: r.bytes(len - 10)? }
        }
        0x19 => {
            let len = r.u32()? as usize;
            let end = r.pos + len;
            let pause = r.u16()?;
            let totp = r.u32()?;
            let npp = r.u8()?;
            let mut asp = r.u8()? as u32;
            let totd = r.u32()?;
            let npd = r.u8()?;
            let mut asd = r.u8()? as u32;
            if asp == 0 {
                asp = 256;
            }
            if asd == 0 {
                asd = 256;
            }
            let mut pilot_symbols: Vec<SymDef> = Vec::new();
            let mut pilot_stream: Vec<PilotRun> = Vec::new();
            if totp > 0 {
                pilot_symbols = read_sym_defs(r, asp as usize, npp)?;
                for _ in 0..totp {
                    let symbol = r.u8()?;
                    let reps = r.u16()?;
                    pilot_stream.push(PilotRun { symbol, reps });
                }
            }
            let mut data_symbols: Vec<SymDef> = Vec::new();
            let mut data: Vec<u8> = Vec::new();
            if totd > 0 {
                data_symbols = read_sym_defs(r, asd as usize, npd)?;
                let bits = u64::from(bits_per_symbol(asd)) * u64::from(totd);
                data = r.bytes(bits.div_ceil(8) as usize)?;
            }
            // Some files are sloppy; trust the declared block length.
            r.pos = end;
            Body::Generalized { pause, totp, npp, pilot_symbols, pilot_stream, totd, npd, data_symbols, data }
        }
        0x20 => Body::Pause { pause: r.u16()? },
        0x21 => {
            let n = r.u8()? as usize;
            Body::GroupStart { name: r.str(n)? }
        }
        0x22 => Body::GroupEnd,
        0x23 => Body::Jump { offset: r.i16()? },
        0x24 => Body::LoopStart { count: r.u16()? },
        0x25 => Body::LoopEnd,
        0x26 => {
            let n = r.u16()?;
            let mut offsets = Vec::with_capacity(n as usize);
            for _ in 0..n {
                offsets.push(r.i16()?);
            }
            Body::Call { offsets }
        }
        0x27 => Body::Return,
        0x28 => {
            let len = r.u16()? as usize;
            let end = r.pos + len;
            let n = r.u8()?;
            let mut entries = Vec::with_capacity(n as usize);
            for _ in 0..n {
                let offset = r.i16()?;
                let l = r.u8()? as usize;
                entries.push(SelectEntry { offset, text: r.str(l)? });
            }
            r.pos = end;
            Body::Select { entries }
        }
        0x2a => {
            r.u32()?; // length, always 0
            Body::Stop48
        }
        0x2b => {
            r.u32()?; // length, always 1
            Body::SignalLevel { level: r.u8()? }
        }
        0x30 => {
            let n = r.u8()? as usize;
            Body::Text { text: r.str(n)? }
        }
        0x31 => {
            let time = r.u8()?;
            let n = r.u8()? as usize;
            Body::Message { time, text: r.str(n)? }
        }
        0x32 => {
            let len = r.u16()? as usize;
            let end = r.pos + len;
            let n = r.u8()?;
            let mut entries = Vec::with_capacity(n as usize);
            for _ in 0..n {
                let kind = r.u8()?;
                let l = r.u8()? as usize;
                entries.push(ArchiveEntry { kind, text: r.str(l)? });
            }
            r.pos = end;
            Body::Archive { entries }
        }
        0x33 => {
            let n = r.u8()?;
            let mut entries = Vec::with_capacity(n as usize);
            for _ in 0..n {
                entries.push(HardwareEntry { kind: r.u8()?, id: r.u8()?, info: r.u8()? });
            }
            Body::Hardware { entries }
        }
        0x35 => {
            let ident = r.str(16)?;
            let len = r.u32()? as usize;
            Body::Custom { ident, data: r.bytes(len)? }
        }
        0x5a => Body::Glue { raw: r.bytes(9)? },
        // Deprecated blocks: keep raw so they round-trip.
        0x16 | 0x17 => {
            let start = r.pos;
            let len = r.u32()? as usize;
            r.pos = start;
            Body::Unknown { id, raw: r.bytes(4 + len)? }
        }
        0x34 => Body::Unknown { id, raw: r.bytes(8)? },
        0x40 => {
            let start = r.pos;
            r.u8()?;
            let len = r.u24()? as usize;
            r.pos = start;
            Body::Unknown { id, raw: r.bytes(4 + len)? }
        }
        // General extension rule: any unknown block has a DWORD length.
        _ => {
            let start = r.pos;
            let len = r.u32()? as usize;
            r.pos = start;
            Body::Unknown { id, raw: r.bytes(4 + len)? }
        }
    })
}

/// Parse a TAP file into standard speed blocks.
pub fn parse_tap(buf: &[u8]) -> ParsedTape {
    let mut r = Reader::new(buf);
    let mut blocks: Vec<Block> = Vec::new();
    let mut warnings: Vec<String> = Vec::new();
    while !r.eof() {
        if r.remaining() < 2 {
            warnings.push("Trailing byte ignored at end of TAP file".to_string());
            break;
        }
        let len = r.u16().unwrap() as usize;
        if len > r.remaining() {
            warnings.push(format!(
                "Truncated TAP block {}: declared {} bytes, {} available",
                blocks.len() + 1,
                len,
                r.remaining()
            ));
            let rest = r.remaining();
            blocks.push(Block::new(Body::Standard { pause: 1000, data: r.bytes(rest).unwrap() }));
            break;
        }
        blocks.push(Block::new(Body::Standard { pause: 1000, data: r.bytes(len).unwrap() }));
    }
    ParsedTape { blocks, major: 1, minor: 20, warnings }
}

/// Auto-detect TZX vs TAP by signature.
pub fn parse_tape(buf: &[u8]) -> Result<ParsedTape, String> {
    if is_tzx(buf) {
        parse_tzx(buf)
    } else {
        Ok(parse_tap(buf))
    }
}
